### PR TITLE
Heretic: colored ammo widget

### LINKED ISSUE #45: Coloured ammo widget
I started working on colourizing the ammo widget, but thought I should ask for your opinion on some questions. Here's a mockup:

HTIC04

Questions:
- Do we want an option for doing this, or is it enough to colour it if "colored elements" for the statusbar is on?
    If we want a separate option, it should probably be split into "ammo+weapons", "ammo only", "weapons only" items, I think. (Here I am colourizing the weapon letter for weapons that the player has).
- Do 0s for 'no ammo' get coloured? They should to be consistent with the way statusbar/hud ammo is I guess.
- Should the ammo totals be colourized as well for the full widget? (Coloring the slash but not the total like I did here looks bad, that can be changed easily though.)
- I would like to add some indication if the player has the gauntlets. Adding a 'g' is easy enough, but then it only makes sense for it to always appear if the weapon letters are coloured. Also, should it go above or below the other weapon/ammo lines?
- Any other thoughts?

### PR DESCRIPTION
**Pull request by kitchen-ace**
Closes #45 

Things for you to check/decide on:

- Line lengths, variable lengths, general formatting. [This line](https://github.com/JNechaevsky/international-doom/commit/8332db94e6cc2b5037f393d1b8d1e3628cad30c8#diff-ebca5aac746b7e31419da0d1099371152b08942d1bb47b7a554cd4e716b7dfa0R3137) is ugly but I don't know your preference for extending the whole table.
- The name "ammo widget coloring" is kind of confusing since the weapon letters are colored even when it's off, but there's only so much room to type. "Extend widget colors" maybe?
- I put in multiple options for coloring the ammo/weapons but maybe you want it just on/off.
- SB_AmmoWidgetColor could be broken into two functions

Thanks!

### CODE REVIEW

**Comment by kitchen-ace** [comment]
Full coloring, full widget:
HTIC00

**Comment by JNechaevsky** [comment]
This is absolutely epic! 😮 I have nothing to add, just one thing to ask - please, feel free to sign your code with your name or nickname, there is no need be so conservative here. 🙂

**Comment by kitchen-ace** [comment]
"Absolutely epic" is overselling it a bit I think, but I'm glad you're happy with it :)

I'm fine with my name just appearing in the commit logs for this, there's not a lot of "new" code, mostly just copy-pasted and changed as necessary.

**Comment by JNechaevsky** [comment]
So be it then, and thank you very much again! I didn't imagine that widget can look so neat while full color.

My next small things in the list, collected while last days. @Meerschweinmann might be interesting to check out as well:
* Disallow player movement while active menu in multiplayer. It's seems to be safe for cross-port compatibility and I did already in Doom. Not needed normally, will be useful for `-solo-net` parameter. But there is also `-coop_spawns` available for Heretic.
* Enabled shadows are not applied to typing string in multiplayer chat. Just a small oversight, easy to fix.
* Check if sound attenuation is correct on extreme distances. While [X/Y/Z](https://github.com/JNechaevsky/international-doom/blob/main/src/heretic/s_sound.c#L173-L175) attenuation distances are using `int64_t` types, final [actual distance](https://github.com/JNechaevsky/international-doom/blob/main/src/heretic/s_sound.c#L168) is still limited to `int`. Sad, but original Heretic doesn't have much verticality, so Z-axis attenuation is almost never notable.
* Playtest more, as my coding/playing ratio is about 95%/5%. It is strange, but improving game code is much more interesting than playing itself.
* Send @rfomin a postcard featuring hand drawn D'Sparil. 👹

**Comment by Meerschweinmann** [comment]
Here is a test map to check the sound-distance. It is one really long sector. You start at 31680 map-units distance from a Maulotaur and a Disciple.
When you shoot at that distance, the Maulotaur's roar can be heard. If you go near to the 1536 map units point the Disciples attack is noticable.
At distance 4096 are numbers with distances on the floor. Hope that is what you want to test.

[H_SNDDST.zip](https://github.com/JNechaevsky/international-doom/files/13961475/H_SNDDST.zip)

**Comment by Meerschweinmann** [comment]
I could edit the map that the long sector does not only go from bottom to top, but also from left to right. And a room with extremely high and deep sitting enemies if that helps you.

**Comment by JNechaevsky** [comment]
No-no, the problem is not about creating a testing map. The problem is I'm at day job now, and despite of being free to use headphones, this still need some calm home ambience. 🙂

**Comment by Meerschweinmann** [comment]
I had tested with a Disciple at 4096 units height and player on 0. No difference to when i stand next to the Disciple. You are correct with Z-axis attenuation is almost never notable.

**Comment by JNechaevsky** [comment]
Huh? You mean, Z-axis is not working in this case, or I got it wrong?

**Comment by Meerschweinmann** [comment]
Yes, there is no difference (or i do not hear a difference)  between standing with the disciple on its 128x128 units platform or being 4096 Units below the Disciple.

**Comment by JNechaevsky** [comment]
But there is difference. 😯 Please take a try this map: [H_SNDDSTv3.zip](https://github.com/JNechaevsky/international-doom/files/13966826/H_SNDDSTv3.zip)

There are two platforms where Disciples are standing, first with `128` units high, second is `4096`. Just wake up one of them by using `iDCLIP` cheat and jump down. 
- When "X/Y/Z" is off, there will be no difference. You can even hear attenuation while he is playing sound and you are jumping down.
- When is on, higher one Disciple becomes too far for attenuation and it will be impossible to head his sounds.

**Comment by Meerschweinmann** [comment]
Oh, sorry. That was my fault. Now i got it. The sfx attenuation axises was on x/y. With x/y/z selected i can clearly hear it. Hope i had not stressed you.

**Comment by JNechaevsky** [comment]
Barely got a heath attack by thinking that feature stop working, but it's alright. 😅

The thing is - both original Doom and Heretic were not taking vertical `Z` axis for sound attenuation, just horizontal `X` and `Y`. So it is never matter how far monster is it height, it can be literally in outer space vertically, but if it's placed near player horizontally, it will sounds like it's near by. Probably such Z-check wasn't done to avoid some extra calculations, probably because there was not much _verticality_. I believe, the first verticality have appeared in Doom II in maps like Downtown, Industrial Zone and Chasm. And probably the very first commercial level with a lot of verticality was a TNT's Deepest Reaches (MAP16) with small open outdoor area and pinkies far below player.

AFAIR, Hexen and Stife were not taking `Z` as well, but things are slightly different in Hexen and I never learned Strife much.

This fix is applying to just to monsters, but to audible map objects as well: 
* Doors (sound emitting from the top sector).
* Lifts/platforms (sound emitting from the bottom sector).
* Lines/switches (sound emitting from the center sector, but this is a bit tricky).

Meanwhile, I made some extra tests and looked at debug prints to make sure that actual distance doesn't get overflow - everything is fine, hooray! Next on the list: review Extra Translucency feature, which making some initially opaque objects translucent. It need addition revision, as some objects get such translucency, while some are not, there is still some potential and possible attention to details.

**Comment by Meerschweinmann** [comment]
That was certainly not my intention to shock you :)

By the way, that is a nice Feature to not irritate the Player with sounds of monsters that are far below or up the Player. For nearly all of the original maps not the big problem, but there are many pwads that use much more verticality.

Since you told me how to build Inter-DOOM from source (23 december), i play Heretic with Inter-Heretic. 

Only one time i got a unreproducable Crash on E2M4 as i stepped into the lava. That was last year and, as i wrote, unreproducable and the only one. But as we say in germany "einmal ist keinmal" that means something like "when it happens only one time, it has never happened".

And since then there were many new Inter-Heretic builds and i never saw this or any other crash with Inter-Heretic. It runs rockstable, smooth and looks beautyful.

**Comment by Meerschweinmann** [comment]
Btw i love that brightmaps.

HTIC00

**Comment by JNechaevsky** [comment]
So it wasn't in illusion with E3M9. Two cases:

1. E3M9 is visited by `IDCLEV` cheat and finished. Mark will stay.
2. E3M9 is visited by entering via secret teleporter on E3M3, and ta-da, cross will dissapear after it's animation: https://drive.google.com/file/d/10zCuU-eQih3v8NQySm2gu2aBDa3JUlgX/view?usp=sharing

This should fix it:
```diff
+++ D:/JN/Programs/MSYS/home/Julia/international-doom/src/heretic/g_game.c	Mon Jan 22 14:28:14 2024
@@ -2095,6 +2095,10 @@
     else if (gamemap == 9)
     {                           // Finished secret level
         gamemap = afterSecret[gameepisode - 1];
+        for (i = 0 ; i < MAXPLAYERS ; i++)
+        {
+            //players[i].didsecret = true;
+        }
     }
     else if (gamemap == 8)
     {

```

**Comment by Meerschweinmann** [comment]
Hi Julia, 

i have seen that you were very busy with Inter-Heretic :) To be honest, i had not noticed the E3M9 cross issue. But very good that you could fix that.

I had noticed too that the ghost enemies seem to be a little less noticable in some situations. But that is not really a problem for me. In my opinion that is a side-effect of 32bit rendering, because the colors of the textures and sprites have lesser rough contrast in colors than with paletted 8bit rendering. The opacity of the monsters is the same when i focus on texture colors that don't change wenn i switch between 32bit and 8bit.

**Comment by JNechaevsky** [comment]
> In my opinion that is a side-effect of 32bit rendering, because the colors of the textures and sprites structures have lesser rough contrast in colors than with paletted 8bit rendering.

That's exactlly the case, indeed! I think I need to increase opacity very-very-very a little bit. Have to carefully experement when eyes will be more fresh. 🙂

> don't change wenn i switch between 32bit and 8bit

Sadly, yes. It is simply impossible to use old tablified translucensy when true color support is compiled in, that's why ghost monsters or any kind of translucensy are always looking decent, matters no if true color is turned on or off in the game menu. 😕 In theory, it is possible to slightly "mess" colors when true color is turned off, but in the end such action will increase translucensy calculations and result still will not be identical to vanilla translucensy.
